File: Tree/BT/993_E_CousinsBinaryTree.js
/**
 * 7.2 night
 * https://leetcode.com/problems/cousins-in-binary-tree/
 */

function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
}

// Accepted --- 116ms 36.7MB 5.19%  reference: https://www.cnblogs.com/seyjs/p/10410669.html
let x_p = null;
let y_p = null;
let x_l = 0;
let y_l = 0;
const isCousins = (root, x, y) => {
    x_p = null;
    y_p = null;
    x_l = 0;
    y_l = 0;
    helper(root, 0, null, x, y)
    return x_p != y_p && x_l == y_l;
};

const helper = (root, level, parent, x, y) => {
    if (root.val == x) {
        x_p = parent;
        x_l = level;
    } else if (root.val == y) {
        y_p = parent;
        y_l = level;
    }
    if (root.left != null) {
        helper(root.left, level + 1, root, x, y);
    }
    if (root.right != null) {
        helper(root.right, level + 1, root, x, y);
    }
};

// Accepted --- 88ms 37.8MB 5.94  reference: https://blog.csdn.net/fuxuemingzhu/article/details/87867902
let map = new Map();
const isCousins_dfs = (root, x, y) => {
    map.clear();
    dfs(root, null, 0);
    let px = map.get(x);
    let py = map.get(y);
    // return px[0].parent.val != py[0].parent.val && px[0].depth == py[0].depth; // no need to use val, lc will report Cannot read property 'val' of null
    return px[0].parent != py[0].parent && px[0].depth == py[0].depth;
};

const dfs = (root, parent, depth) => {
    if (!root) return;
    let data = [];
    data.push({
        parent: parent,
        depth: depth,
    })
    map.set(root.val, data);
    dfs(root.left, root, depth + 1);
    dfs(root.right, root, depth + 1);
};

// Accepted --- 100ms 38.7MB 5.19%  reference: https://blog.csdn.net/fuxuemingzhu/article/details/87867902
const isCousins_bfs_queue = (root, x, y) => {
    let q = [];
    q.push({
        first: root,
        second: null
    });
    let map = new Map();
    let depth = 0;
    while (q.length != 0) {
        let n = q.length;
        for (let i = 0; i < n; ++i) {
            let p = q[0];
            q.shift();
            if (!p.first) continue;
            map.set(p.first.val, {
                parent: p.second,
                depth: depth
            });
            q.push({
                first: p.first.left,
                second: p.first
            });
            q.push({
                first: p.first.right,
                second: p.first
            });
        }
        ++depth;
    }
    let px = map.get(x);
    let py = map.get(y);
    return px.parent != py.parent && px.depth == py.depth;
};

const main = () => {
    let root = new TreeNode(1);
    root.left = new TreeNode(2);
    root.right = new TreeNode(3);
    root.left.left = new TreeNode(4);
    let x = 4,
        y = 3;

    let root2 = new TreeNode(1);
    root2.left = new TreeNode(2);
    root2.right = new TreeNode(3);
    root2.left.right = new TreeNode(4);
    root2.right.right = new TreeNode(5);
    let x2 = 5,
        y2 = 4;

    let root3 = new TreeNode(1);
    root3.left = new TreeNode(2);
    root3.right = new TreeNode(3);
    root3.left.right = new TreeNode(4);
    let x3 = 2,
        y3 = 3;

    console.log(isCousins(root, x, y));
    console.log(isCousins(root2, x2, y2));
    console.log(isCousins(root3, x3, y3));

    console.log("");
    console.log(isCousins_dfs(root, x, y));
    console.log(isCousins_dfs(root2, x2, y2));
    console.log(isCousins_dfs(root3, x3, y3));

    console.log("");
    console.log(isCousins_bfs_queue(root, x, y));
    console.log(isCousins_bfs_queue(root2, x2, y2));
    console.log(isCousins_bfs_queue(root3, x3, y3));

};

main()




// const getDepth = (root, val, depth, level) => {
//     if (!root) return null;
//     if ((root.left && root.left.val == val) || (root.right && root.right.val == val)) {
//         level = depth;
//         return root;
//     }
//     let left = getDepth(root.left, val, depth + 1, level);
//     console.log(level); // issue
//     if (left) return left;

//     let right = getDepth(root.right, val, depth + 1, level);
//     if (right) return right;
//     return null;
// };

// const isCousins = (root, x, y) => {
//     let xDepth = -1;
//     let yDepth = -1;
//     let xParent = getDepth(root, x, 0, xDepth);
//     let yParent = getDepth(root, y, 0, yDepth);
//     console.log(xDepth, yDepth, xParent.val, yParent.val);
//     if (xDepth == yDepth && xParent != yParent) {
//         return true;
//     }
//     return false;
// };

File: Tree/BT/993_E_CousinsBinaryTree.cxx
// 7.2 night
#include <iostream>
#include <unordered_map>
#include <queue>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    // https://www.cnblogs.com/douzujun/p/10922786.html
    bool isCousins(TreeNode *root, int x, int y)
    {
        int x_depth = -1, y_depth = -1;
        TreeNode *x_parent = getDepth(root, x, 0, x_depth);
        TreeNode *y_parent = getDepth(root, y, 0, y_depth);
        // cout << x_depth << " " << y_depth << " " << x_parent->val << " " << y_parent->val << " " << endl;
        if (x_depth == y_depth && x_parent != y_parent)
        {
            return true;
        }
        return false;
    }

    TreeNode *getDepth(TreeNode *root, int val, int depth, int &level)
    {
        if (!root)
        {
            return nullptr;
        }

        if ((root->left && root->left->val == val) || (root->right && root->right->val == val))
        {
            level = depth;
            return root;
        }
        TreeNode *left = getDepth(root->left, val, depth + 1, level);
        // cout << level << endl;
        if (left)
        {
            return left;
        }

        TreeNode *right = getDepth(root->right, val, depth + 1, level);
        if (right)
        {
            return right;
        }

        return nullptr;
    }

    // https://blog.csdn.net/fuxuemingzhu/article/details/87867902
    unordered_map<int, pair<TreeNode *, int>> m_;
    bool isCousins2(TreeNode *root, int x, int y)
    {
        m_.clear();
        dfs(root, nullptr, 0);
        auto px = m_[x], py = m_[y];
        // cout << px.first->val << " " << py.first->val << endl;
        // cout << px.second << " " << py.second << endl;
        return px.first != py.first && px.second == py.second;
    }

    void dfs(TreeNode *root, TreeNode *parent, int depth)
    {
        if (!root)
            return;
        m_[root->val] = make_pair(parent, depth);
        dfs(root->left, root, depth + 1);
        dfs(root->right, root, depth + 1);
    }
};

int main()
{
    Solution s;

    TreeNode *root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    int x = 4, y = 3;

    TreeNode *root2 = new TreeNode(1);
    root2->left = new TreeNode(2);
    root2->right = new TreeNode(3);
    root2->left->right = new TreeNode(4);
    root2->right->right = new TreeNode(5);
    int x2 = 5, y2 = 4;

    TreeNode *root3 = new TreeNode(1);
    root3->left = new TreeNode(2);
    root3->right = new TreeNode(3);
    root3->left->right = new TreeNode(4);
    int x3 = 2, y3 = 3;

    cout << s.isCousins(root, x, y) << endl;
    cout << s.isCousins(root2, x2, y2) << endl;
    cout << s.isCousins(root3, x3, y3) << endl;

    cout << endl;
    cout << s.isCousins2(root, x, y) << endl;
    cout << s.isCousins2(root2, x2, y2) << endl;
    cout << s.isCousins2(root3, x3, y3) << endl;
}
